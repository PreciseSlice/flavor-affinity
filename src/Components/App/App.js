import React, { Component } from 'react';
import { connect } from 'react-redux';
import './App.css';
import MainForm from '../MainForm/MainForm';
import CardContainer from '../CardContainer/CardContainer';
//import { getAllIngredients } from '../../Helpers/apiCalls';
import { setAllIngredients } from '../../Actions';
//import * as actions from '../../Actions';
import { allDataClean } from '../../Helpers/.mockApi.js';
//import { cleanData } from '../App/testData';

export class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      error: null,
      errorInfo: null
    };
  }

  async componentDidMount() {
    const { setAllIngredients } = this.props;
    //const allIngredients = await getAllIngredients();
    //setAllIngredients(allIngredients);
    setAllIngredients(allDataClean);
  }

  componentDidCatch(error, errorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo
    });
  }

  clearError() {
    this.setState({
      error: null,
      errorInfo: null
    });
  }

  render() {
    if (this.state.error) {
      return (
        <div className="eror">
          <h2>{'Oh-no! Something went wrong'}</h2>
          <p className="">{this.state.error && this.state.error.toString()}</p>
          <div>{'Component Stack Error Details: '}</div>
          <p className="">{this.state.errorInfo.componentStack}</p>
          <button onClick={() => this.clearError()} />
        </div>
      );
    }
    return (
      <div className="app">
        <MainForm />
        <CardContainer />
      </div>
    );
  }
}

// export const mapStateToProps = state => ({

// });

export const mapDispatchToProps = dispatch => ({
  setAllIngredients: ingredient => dispatch(setAllIngredients(ingredient))
});

export default connect(null, mapDispatchToProps)(App);
