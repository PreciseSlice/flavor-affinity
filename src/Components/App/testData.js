/* eslint-disable */
export const allIngredients = [
  {
    id: 5902,
    name: '1724 Tonic',
    product: '1724 Tonic',
    preparation: '',
    _links: {
      preparation: null,
      origin: null,
      image: {
        id: 4578,
        src: 'https://pantry.foodpairing.com/apps/4578',
        size_240: 'https://pantry.foodpairing.com/apps/4578_240_0',
        size_480: 'https://pantry.foodpairing.com/apps/4578_480_0',
        size_720: 'https://pantry.foodpairing.com/apps/4578_720_0',
        size_960: 'https://pantry.foodpairing.com/apps/4578_960_0',
        size_1200: 'https://pantry.foodpairing.com/apps/4578_1200_0',
        size_1440: 'https://pantry.foodpairing.com/apps/4578_1440_0'
      },
      brand: null,
      categories: [
        {
          id: 38,
          href:
            'https://api.foodpairing.com/taxonomies/ingredients/categories/38'
        },
        {
          id: 139,
          href:
            'https://api.foodpairing.com/taxonomies/ingredients/categories/139'
        }
      ]
    },
    description:
      'The essence of 1724 Tonic Water has its origins in the Andes, on the mythical Inca Trail. This is not by chance. This is where quinine was discovered. There, this ingredient gains the authenticity of a product with roots in the region that go back hundreds of years. At 1724 metres above sea level. Not a metre above or below.',
    _meta: {
      created_at: '',
      modified_at: '',
      index: 0
    }
  },
  {
    id: 5440,
    name: 'A la Chimay Bleue',
    product: 'A la Chimay Bleue',
    preparation: '',
    _links: {
      preparation: null,
      origin: null,
      image: {
        id: 1772,
        src: 'https://pantry.foodpairing.com/apps/1772',
        size_240: 'https://pantry.foodpairing.com/apps/1772_240_0',
        size_480: 'https://pantry.foodpairing.com/apps/1772_480_0',
        size_720: 'https://pantry.foodpairing.com/apps/1772_720_0',
        size_960: 'https://pantry.foodpairing.com/apps/1772_960_0',
        size_1200: 'https://pantry.foodpairing.com/apps/1772_1200_0',
        size_1440: 'https://pantry.foodpairing.com/apps/1772_1440_0'
      },
      brand: {
        id: 80,
        name: 'Chimay',
        website: '',
        description: '',
        _links: {
          images: [
            {
              id: 4420,
              src: 'https://pantry.foodpairing.com/apps/4420'
            }
          ]
        }
      },
      categories: [
        {
          id: 265,
          href:
            'https://api.foodpairing.com/taxonomies/ingredients/categories/265'
        },
        {
          id: 266,
          href:
            'https://api.foodpairing.com/taxonomies/ingredients/categories/266'
        }
      ]
    },
    description:
      'The rind of the à la Chimay Bleue is regularly washed with the Chimay Blue Cap during his stay in the aging cellars of the monestarey. Pronounced flavor with a slight bitterness of the hops.',
    _meta: {
      created_at: '',
      modified_at: '',
      index: 1
    }
  },
  {
    id: 5443,
    name: 'A la Chimay Rouge',
    product: 'A la Chimay Rouge',
    preparation: '',
    _links: {
      preparation: null,
      origin: null,
      image: {
        id: 1773,
        src: 'https://pantry.foodpairing.com/apps/1773',
        size_240: 'https://pantry.foodpairing.com/apps/1773_240_0',
        size_480: 'https://pantry.foodpairing.com/apps/1773_480_0',
        size_720: 'https://pantry.foodpairing.com/apps/1773_720_0',
        size_960: 'https://pantry.foodpairing.com/apps/1773_960_0',
        size_1200: 'https://pantry.foodpairing.com/apps/1773_1200_0',
        size_1440: 'https://pantry.foodpairing.com/apps/1773_1440_0'
      },
      brand: {
        id: 80,
        name: 'Chimay',
        website: '',
        description: '',
        _links: {
          images: [
            {
              id: 4420,
              src: 'https://pantry.foodpairing.com/apps/4420'
            }
          ]
        }
      },
      categories: [
        {
          id: 265,
          href:
            'https://api.foodpairing.com/taxonomies/ingredients/categories/265'
        },
        {
          id: 266,
          href:
            'https://api.foodpairing.com/taxonomies/ingredients/categories/266'
        }
      ]
    },
    description:
      'The à la Chimay Rouge is a round, smooth cheese with creamy texture. Its natural rind is washed with Chimay Red Cap. Fruity flavour with slight bitterness.',
    _meta: {
      created_at: '',
      modified_at: '',
      index: 2
    }
  }
];

export const cleanData = [
    {
      id: 5902,
      name: '1724 tonic',
      image: 'https://pantry.foodpairing.com/apps/4578_240_0',
      description:
        'The essence of 1724 Tonic Water has its origins in the Andes, on the mythical Inca Trail. This is not by chance. This is where quinine was discovered. There, this ingredient gains the authenticity of a product with roots in the region that go back hundreds of years. At 1724 metres above sea level. Not a metre above or below.'
    },
    {
      id: 5440,
      name: 'a la chimay bleue',
      image: 'https://pantry.foodpairing.com/apps/1772_240_0',
      description:
        'The rind of the à la Chimay Bleue is regularly washed with the Chimay Blue Cap during his stay in the aging cellars of the monestarey. Pronounced flavor with a slight bitterness of the hops.'
    },
    {
      id: 5443,
      name: 'a la chimay rouge',
      image: 'https://pantry.foodpairing.com/apps/1773_240_0',
      description:
        'The à la Chimay Rouge is a round, smooth cheese with creamy texture. Its natural rind is washed with Chimay Red Cap. Fruity flavour with slight bitterness.'
    }
];
