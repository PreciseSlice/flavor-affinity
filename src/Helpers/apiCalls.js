import { AppId, AppKey } from './.apiKey.js';

export const fetchFromApi = async url => {
    const initialFetch = await fetch(url, {
      headers: {
        'X-Application-ID': AppId,
        'X-Application-Key': AppKey
      }
    });

  return await validateResponce(initialFetch)
};

export const validateResponce = initialFetch => {
  if (initialFetch.status <= 200) {
    return initialFetch.json();
  } else {
    throw new Error('Status code > 200');
  }
}

export const searchForIngredient = async ingredient => {
  const searchResult = await fetchFromApi(
    `https://api.foodpairing.com/ingredients/?q=${ingredient}`
  );

  return searchResult;
};

export const getAllIngredients = async () => {
  const allIngredients = await fetchFromApi(
    'https://api.foodpairing.com/ingredients/?limit=500'
  );

  return cleanAllIngredients(allIngredients);
};

export const cleanAllIngredients = ingredientsData => {
  return ingredientsData.map(ingredient => {
    return {
      id: ingredient.id,
      name: ingredient.name.toLowerCase(),
      image: ingredient._links.image.size_240,
      description: ingredient.description,
      selected: false
    };
  });
};

export const getParings = async (id, name) => {
  const pairings = await fetchFromApi(
    `https://api.foodpairing.com/ingredients/${id}/pairings`
  );

  const allPairings = cleanPairings(pairings);

  return slicePairings(allPairings, name);
};

export const cleanPairings = ingredientsData => {
  return ingredientsData.map(ingredient => {
    return {
      id: ingredient._links.ingredient.id,
      name: ingredient._links.ingredient.name.toLowerCase(),
      image: ingredient._links.ingredient._links.image.size_240
    };
  });
};

export const slicePairings = (allPairings, name) => {
  const topFive = allPairings.slice(0, 5);
  const middleFive = allPairings.slice(5, 10);
  const finalFive = allPairings.slice(10, 15);

  return {
    name,
    topFive,
    middleFive,
    finalFive
  };
};
