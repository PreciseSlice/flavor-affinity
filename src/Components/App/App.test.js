/* eslint-disable */
import React from 'react';
import ReactDOM from 'react-dom';
import { App, mapDispatchToProps } from './App';
import { shallow, mount } from 'enzyme';
import { allIngredients } from './testData';
import { getAllIngredients } from '../../Helpers/apiCalls';

window.fetch = jest.fn().mockImplementation(url => {
  return Promise.resolve({
    status: 200,
    json: () => Promise.resolve(allIngredients)
  });
});

let mockFn;

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    mockFn = jest.fn();
    wrapper = shallow(<App setIngredients={mockFn} />);
  });

  it('Exist and matches snapshot', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default states of error and errorInfo set to null', () => {
    expect(wrapper.state().error).toEqual(null);
    expect(wrapper.state().errorInfo).toEqual(null);
  });

  describe('componentDidMount', () => {
    it('calls componentDidMount() lifecycle method', () => {
      const spy = jest.spyOn(App.prototype, 'componentDidMount');
      const wrapper = mount(<App setIngredients={mockFn} />);

      expect(App.prototype.componentDidMount).toHaveBeenCalled();

      spy.mockReset();
      spy.mockRestore();
    });

    it.skip('calls getAll ingredients and setIngredients when component mounts', () => {
      expect(wrapper.instance().props.setIngredients).toHaveBeenCalled();
      //expect(getAllIngredients).toHaveBeenCalled()
    });
  });

  describe.skip('componentDidCatch', () => {
    const ProblemChild = () => {
      throw new Error('Error thrown from problem child');
      return <div>Error</div>;
    };

    it('should catch errors with componentDidCatch', () => {
      console.error = mockFn;
      const spy = jest.spyOn(App.prototype, 'componentDidCatch');
      const wrapper = mount(
        <App setIngredients={mockFn}>
          <ProblemChild />
        </App>
      );
      expect(App.prototype.componentDidCatch).toHaveBeenCalled();

      spy.mockReset();
      spy.mockRestore();
    });
  });

  describe('clearError', () => {
    it('should clear error by setting the state of error and errorInfo back to null', () => {
      expect(wrapper.state().error).toEqual(null);
      expect(wrapper.state().errorInfo).toEqual(null);

      wrapper.setState({
        error: 500,
        errorInfo: 'shits on fire yo'
      });

      expect(wrapper.state().error).toEqual(500);
      expect(wrapper.state().errorInfo).toEqual('shits on fire yo');

      wrapper.instance().clearError();

      expect(wrapper.state().error).toEqual(null);
      expect(wrapper.state().errorInfo).toEqual(null);
    });
  });
});
