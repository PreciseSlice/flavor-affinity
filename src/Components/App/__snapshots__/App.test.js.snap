// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Exist and matches snapshot 1`] = `
<div
  className="App"
>
  <Connect(MainForm) />
</div>
`;
